File: c++/lecture/class_test/account/service/AccountServiceImpl.cpp
//
// Created by eddi on 23. 12. 5.
//

#include "AccountServiceImpl.h"
#include "response/AccountRegisterResponse.h"

AccountServiceImpl::AccountServiceImpl(std::shared_ptr<AccountRepository> accountRepository) : accountRepository(accountRepository) { }

AccountRegisterResponse *
        AccountServiceImpl::create(AccountRegisterRequest *request)
{
    if (accountRepository->getBoolWithFindByAccountId(request->getAccountId())) {
        return nullptr;
    }

    Account *account = accountRepository->save(request->toAccount());

    if (account != nullptr) {
        return new AccountRegisterResponse(true);
    }

    return nullptr;
}

AccountLoginResponse *
AccountServiceImpl::signIn(AccountLoginRequest *request)
{
    std::optional<Account> accountOpt = accountRepository->findByAccountId(request->getAccountId());

    if (accountOpt.has_value()) {
        std::cout << "매칭되는 사용자를 찾았습니다" << std::endl;
        Account account = accountOpt.value();

        if (account.get_password() == request->getPassword()) {
            return new AccountLoginResponse(true);
        } else {
            std::cout << "사용자 인증에 실패하였습니다" << std::endl;
            return nullptr;
        }
    } else {
        std::cout << "사용자 계정을 찾지 못했습니다" << std::endl;
        return nullptr;
    }

    return nullptr;
}

void AccountServiceImpl::signOut(int sessionId)
{
    // 사실 이러면 안되지만 적당히 퉁칩시다.
    accountRepository->deleteSession(sessionId);
}

AccountServiceImpl& AccountServiceImpl::getInstance(
        std::shared_ptr<AccountRepository> accountRepository)
{
    static AccountServiceImpl instance(accountRepository);
    return instance;
}

File: python/lecture/answerProcess/account/service/AccountServiceImpl.py
from account.entity.Session import Session
from account.repository.AccountRepositoryImpl import AccountRepositoryImpl
from account.repository.SessionRepositoryImpl import SessionRepositoryImpl
from account.service.AccountService import AccountService
from account.service.request.AccountDeleteRequest import AccountDeleteRequest
from account.service.request.AccountLoginRequest import AccountLoginRequest
from account.service.request.AccountLogoutRequest import AccountLogoutRequest
from account.service.request.AccountRegisterRequest import AccountRegisterRequest
from account.service.response.AccountDeleteResponse import AccountDeleteResponse
from account.service.response.AccountLoginResponse import AccountLoginResponse
from account.service.response.AccountLogoutResponse import AccountLogoutResponse
from account.service.response.AccountRegisterResponse import AccountRegisterResponse


class AccountServiceImpl(AccountService):
    __instance = None

    def __new__(cls, accountRepository, sessionRepository):
        if cls.__instance is None:
            cls.__instance = super().__new__(cls)
            cls.__instance.__accountRepository = accountRepository
            cls.__instance.__sessionRepository = sessionRepository
        return cls.__instance
    #
    # def __init__(self):
    #     print("AccountServiceImpl 생성자 호출")

    @classmethod
    def getInstance(cls, repository=None):
        if cls.__instance is None:
            cls.__instance = cls()
        return cls.__instance

    def registerAccount(self, *args, **kwargs):
        print("registerAccount()")
        print(f"args: {args}")

        cleanedElements = args[0]
        print(f"cleanedElements: {cleanedElements}")

        # for i, element in enumerate(cleanedElements):
        #     print(f"각각의 요소 {i + 1}: {element}")

        accountRegisterRequest = AccountRegisterRequest(*cleanedElements)
        accountId = accountRegisterRequest.getAccountId()
        foundAccount = self.__accountRepository.findByAccountId(accountId)
        if foundAccount is not None:
            return AccountRegisterResponse(False)

        storedAccount = self.__accountRepository.save(accountRegisterRequest.toAccount())

        if storedAccount.getId() is not None:
            return AccountRegisterResponse(True)

        return AccountRegisterResponse(False)

    def loginAccount(self, *args, **kwargs):
        print("loginAccount()")
        print(f"args: {args}")

        cleanedElements = args[0]
        print(f"cleanedElements: {cleanedElements}")

        accountLoginRequest = AccountLoginRequest(*cleanedElements)
        foundAccount = self.__accountRepository.findByAccountId(accountLoginRequest.getAccountId())
        print(f"foundAccount: {foundAccount}")
        if foundAccount is None:
            return AccountLoginResponse(-1)

        if foundAccount.checkPassword(accountLoginRequest.getPassword()):
            # sessionRepository = SessionRepositoryImpl.getInstance()
            accountSession = Session(foundAccount.getId())
            # sessionRepository.save(accountSession)
            self.__sessionRepository.save(accountSession)

            return AccountLoginResponse(foundAccount.getId())

        return AccountLoginResponse(-1)

    def logoutAccount(self, *args, **kwargs):
        print("AccountService - logoutAccount()")
        print(f"args: {args}")

        cleanedElements = args[0]
        print(f"cleanedElements: {cleanedElements}")

        accountLoginRequest = AccountLogoutRequest(*cleanedElements)
        foundAccount = self.__accountRepository.findById(accountLoginRequest.getAccountSessionId())
        print(f"foundAccount: {foundAccount}")
        if foundAccount is None:
            return AccountLogoutResponse(False)

        self.__sessionRepository.deleteBySessionId(foundAccount.getId())

        return AccountLogoutResponse(True)

    def deleteAccount(self, *args, **kwargs):
        print("AccountService - deleteAccount()")

        cleanedElements = args[0]

        accountLoginRequest = AccountDeleteRequest(*cleanedElements)
        foundAccount = self.__accountRepository.findById(accountLoginRequest.getAccountSessionId())
        print(f"foundAccount: {foundAccount}")
        if foundAccount is None:
            return AccountDeleteResponse(False)

        self.__sessionRepository.deleteBySessionId(foundAccount.getId())
        self.__accountRepository.deleteById(foundAccount.getId())

        return AccountDeleteResponse(True)


    